Initialize klog with (go)flags and add these flags to pflag as well

jerolimov opened the pull request:
# Changes

Just initialize [klog](https://github.com/kubernetes/klog) with go flags and add these flags to pflag as well.

Fixes #7

With this change its possible to enable k8s api communication, for example:

```bash
go run cmd/shp/main.go -v=10 br list

I1029 19:15:17.183378  141455 loader.go:379] Config loaded from file: ...kubeconfig
I1029 19:15:17.184151  141455 round_trippers.go:425] curl -k -v -XGET  -H "Authorization: Bearer <masked>" -H "Accept: application/json, */*" -H "User-Agent: main/v0.0.0 (linux/amd64) kubernetes/$Format" 'https://api.crc.testing:6443/apis/shipwright.io/v1alpha1/namespaces/default/buildruns'
I1029 19:15:17.195693  141455 round_trippers.go:445] GET https://api.crc.testing:6443/apis/shipwright.io/v1alpha1/namespaces/default/buildruns 200 OK in 11 milliseconds

...

NAME                    STATUS
s2i-nodejs-build-rjtz9  Failed
```

Help output shows `klog` attributes as well as all subcommands. See `-v` and all these `--log*` arguments.

```
go run cmd/shp/main.go --help

Command-line client for Shipwright's Build API.

Usage:
  shp [command] [resource] [flags]
  shp [command]

Available Commands:
  build       Manage Builds
  buildrun    Manage BuildRuns
  completion  generate the autocompletion script for the specified shell
  help        Help about any command

Flags:
      --add_dir_header                   If true, adds the file directory to the header of the log messages
      --alsologtostderr                  log to standard error as well as files
      --as string                        Username to impersonate for the operation
      --as-group stringArray             Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
      --cache-dir string                 Default cache directory (default "/home/christoph/.kube/cache")
      --certificate-authority string     Path to a cert file for the certificate authority
      --client-certificate string        Path to a client certificate file for TLS
      --client-key string                Path to a client key file for TLS
      --cluster string                   The name of the kubeconfig cluster to use
      --context string                   The name of the kubeconfig context to use
  -h, --help                             help for shp
      --insecure-skip-tls-verify         If true, the server's certificate will not be checked for validity. This will make your HTTPS connections insecure
      --kubeconfig string                Path to the kubeconfig file to use for CLI requests.
      --log_backtrace_at traceLocation   when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                   If non-empty, write log files in this directory
      --log_file string                  If non-empty, use this log file
      --log_file_max_size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --logtostderr                      log to standard error instead of files (default true)
  -n, --namespace string                 If present, the namespace scope for this CLI request
      --one_output                       If true, only write logs to their native severity level (vs also writing to each lower severity level
      --request-timeout string           The length of time to wait before giving up on a single server request. Non-zero values should contain a corresponding time unit (e.g. 1s, 2m, 3h). A value of zero means don't timeout requests. (default "0")
  -s, --server string                    The address and port of the Kubernetes API server
      --skip_headers                     If true, avoid header prefixes in the log messages
      --skip_log_headers                 If true, avoid headers when opening log files
      --stderrthreshold severity         logs at or above this threshold go to stderr (default 2)
      --tls-server-name string           Server name to use for server certificate validation. If it is not provided, the hostname used to contact the server is used
      --token string                     Bearer token for authentication to the API server
      --user string                      The name of the kubeconfig user to use
  -v, --v Level                          number for the log level verbosity
      --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging

Use "shp [command] --help" for more information about a command.
```

# Submitter Checklist

- [x] Includes tests if functionality changed/was added
- [ ] Includes docs if changes are user-facing
- [x] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [x] Release notes block has been filled in, or marked NONE

See [the contributor guide](https://github.com/shipwright-io/build/blob/main/CONTRIBUTING.md)
for details on coding conventions, github and prow interactions, and the code review process.

# Release Notes

```release-note
Initialize klog, this enable users to customize k8s api logging with -v and --log* arguments.
```

adambkaplan commented on cmd/shp/main.go:
```diff
@@ -15,8 +17,8 @@ import (
 func main() {
-	flags := pflag.NewFlagSet(ApplicationName, pflag.ExitOnError)
-	pflag.CommandLine = flags
+	init_goflags()
+	init_pflags()
```
Style nit - go functions typically don't use underscores in names. This is more of a convention thing than anything enforced in the language:
```suggestion
	initGoFlags()
	initPFlags()
```

adambkaplan commented on cmd/shp/main.go:
```diff
@@ -25,3 +27,24 @@ func main() {
 	}
 }
+
+// goflags ignores -h and --help because pflag will show the usage later with all subcommands.
+func init_goflags() {
```
Go has a handy feature called `godoc` where comments in front of function names automatically generate documentation. For this to work, the first word of the comment and the function name should match. Many linters catch this (and in the future we will probably add these lint checks to our CI).

```suggestion
// initGoFlags initializes the flag sets for klog. Any flags for "-h" or "--help" are ignored because pflag will show the usage later with all subcommands.
func initGoFlags() {
```

adambkaplan commented on cmd/shp/main.go:
```diff
@@ -25,3 +27,24 @@ func main() {
+	flagset.Parse(args)
+}
+
+func init_pflags() {
```
Style nit:
```suggestion
// initPFlags initializes the pflags used by Cobra subcommands.
func initPFlags() {
```

adambkaplan reviewed (changes requested):
@jerolimov and I did a live coding session where we experimented with other kubernetes libraries to initialize klog's flags. Unfortunately this didn't work - the functions here are the best we can do. I have added a few stylistic suggestions to make the code more "go-like".

@jerolimov I think you can also add an e2e test to verify we are adding the desired flags. We use the [Bats](https://github.com/bats-core/bats-core) framework for running the `shp` command line as if it were run by an end user. You can an example of this in the [run-follow](https://github.com/shipwright-io/cli/blob/main/test/e2e/run-follow.bats) test. The bats assert libraries will probably help as well.

adambkaplan commented:
/kind feature

jerolimov commented:
Thanks @adambkaplan for the feedback. I've updated the PR with "go style" function names and godocs, and rebased it as well.

I've also added some bat e2e tests for the `--help` output in general and esp. for the `-v=10` case.

Please take a new look. Thanks. :slightly_smiling_face: 

The last envvars.bats tests fails for me on main and with this PR, so I comment out the last 3 asserts in envvars.bats (only locally).

```bash
➜  cli git:(init-klog) ✗ make build test-e2e
go build -v -mod=vendor -o _output/shp ./cmd/shp/...
./test/e2e/bats/core/bin/bats --recursive test/e2e/*.bats
 ✓ shp binary can be executed
 ✓ shp --help lists all available commands
 ✓ shp --help lists some common flags
 ✓ shp --help lists also logging flags
 ✓ shp [build/buildrun] create should not error when a name is specified
 ✓ shp -v=10 build list can log the kubernetes api communication
 ✓ shp environment variable lifecycle
 ✓ shp build run follow verification

8 tests, 0 failures
```

/cc @coreydaley this PR doesn't update any output from the `shp`. So the list output (feature request: it should show the states in colors!) or errors doesn't add any timestamps. Only if you enable additional logging it shows this output with the timestamps. Sorry that I didn't picked up your earlier work, didn't know that PR on the hack event on Friday.

adambkaplan reviewed (approved):
/approve

Looks good to me.

As an FYI, there has been improvement upstream in the latest version of k8s.io/component-base that simplifies initialization of klog.
We should consider refactoring to it with our next kube rebase.

See https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/cli/run.go

jerolimov commented:
@adambkaplan how do we get this merged? :)

adambkaplan commented:
/assign @otaviof

otaviof reviewed (approved):
Good additions to the CLI, @jerolimov! Thanks.

/lgtm
